packages/ts-client/src/webrtc/transciever.ts

```diff
@@ -0,0 +1,210 @@
+    const typeTransceiversNumber = recvTransceivers.filter(
+      (elem) => elem.receiver.track.kind === type,
+    ).length;
+    return Array(typeNumber - typeTransceiversNumber).fill(type);
+  };
+
+  const audio = getNeededTransceiversTypes('audio');
+  const video = getNeededTransceiversTypes('video');
+  toAdd = toAdd.concat(audio);
+  toAdd = toAdd.concat(video);
+
+  for (const kind of toAdd)
+    connection?.addTransceiver(kind, { direction: 'recvonly' });
+};
+
+export const setTransceiverDirection = (connection: RTCPeerConnection) => {
+  connection
+    .getTransceivers()
+    .forEach(
+      (transceiver) =>
+        (transceiver.direction =
+          transceiver.direction === 'sendrecv'
+            ? 'sendonly'
+            : transceiver.direction),
+    );
+};
+
+export const addTrackToConnection = <EndpointMetadata, TrackMetadata>(
+  trackContext: TrackContext<EndpointMetadata, TrackMetadata>,
+  disabledTrackEncodingsMap: Map<string, TrackEncoding[]>,
+  connection: RTCPeerConnection | undefined,
+) => {
+  const transceiverConfig = createTransceiverConfig(
+    trackContext,
+    disabledTrackEncodingsMap,
+  );
+  const track = trackContext.track!;
+  connection!.addTransceiver(track, transceiverConfig);
+};
+
+const createTransceiverConfig = <EndpointMetadata, TrackMetadata>(
+  trackContext: TrackContext<EndpointMetadata, TrackMetadata>,
+  disabledTrackEncodingsMap: Map<string, TrackEncoding[]>,
+): RTCRtpTransceiverInit => {
+  let transceiverConfig: RTCRtpTransceiverInit;
+
+  if (trackContext.track!.kind === 'audio') {
+    transceiverConfig = createAudioTransceiverConfig(trackContext);
+  } else {
+    transceiverConfig = createVideoTransceiverConfig(
+      trackContext,
+      disabledTrackEncodingsMap,
+    );
+  }
+
+  return transceiverConfig;
+};
+
+const createAudioTransceiverConfig = <EndpointMetadata, TrackMetadata>(
+  trackContext: TrackContext<EndpointMetadata, TrackMetadata>,
+): RTCRtpTransceiverInit => {
+  return {
+    direction: 'sendonly',
+    streams: trackContext.stream ? [trackContext.stream] : [],
+  };
+};
+
+const createVideoTransceiverConfig = <EndpointMetadata, TrackMetadata>(
+  trackContext: TrackContext<EndpointMetadata, TrackMetadata>,
+  disabledTrackEncodingsMap: Map<string, TrackEncoding[]>,
+): RTCRtpTransceiverInit => {
+  let transceiverConfig: RTCRtpTransceiverInit;
+  if (trackContext.simulcastConfig!.enabled) {
+    transceiverConfig = simulcastTransceiverConfig;
+    const trackActiveEncodings = trackContext.simulcastConfig!.activeEncodings;
+    const disabledTrackEncodings: TrackEncoding[] = [];
+    transceiverConfig.sendEncodings?.forEach((encoding) => {
+      if (trackActiveEncodings.includes(encoding.rid! as TrackEncoding)) {
+        encoding.active = true;
+      } else {
+        disabledTrackEncodings.push(encoding.rid! as TrackEncoding);
+      }
+    });
+    disabledTrackEncodingsMap.set(trackContext.trackId, disabledTrackEncodings);
```

Comment:
we mutate a map passed in as an argument and return a completely other value.
please create another function to handle updating the disabled encodings

i can see its being drilled through at least 3 functions, i still think it should get refactored though

Reply:
I will do it in future PRs